### PR TITLE
Added moderation capabilities

### PR DESCRIPTION
**Pull request by Dalethium**
Added the ability to approve or deny posts

### CODE REVIEW

**Comment by Dalethium** [comment]
@mikegcox hired me to code this, just thought I'd give him credit for getting it done :)

**Comment by rhukster** [comment]
Very nice :).

**Comment by flaviocopes** [comment]
Cool! Can you change tab indentation to 4 spaces, as per http://www.php-fig.org/psr/psr-2/? Also remove 

```
error_reporting(E_ALL);
 ini_set('display_errors', 1);
```

**Comment by Dalethium** [comment]
I am used to K&R formatting, my bad. I removed E_all and such, is there anything else that needs done?

**Comment by mikegcox** [comment]
Thanks for the additions to this code @Dalethium.  Hopefully the code can help the community.

File: guestbook.php [hunk 1]
```diff
@@ -131,12 +335,27 @@ public function onTwigTemplatePaths()
...
+        $this->grav['twig']->twig_paths[] = __DIR__ . '/admin/templates';
+    }
 
     /**
      * Exclude comments from the Data Manager plugin
      */
     public function onDataTypeExcludeFromDataManagerPluginHook()
     {
-        $this->grav['admin']->dataTypesExcludedFromDataManagerPlugin[] = 'guestbook';
+        $this->grav['admin']->dataTypesExcludedFromDataManagerPlugin[] = 'a-guestbook';
```
Comment (flaviocopes) [inline guestbook.php]: Why this change?

File: admin/templates/guestbook.html.twig [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                    if (totalRetrieved >= response.totalAvailable) {
+                        $('.totalRetrieved').html($('.totalAvailable').html());
+                        $('.js__load-more').hide();
+                    }
+                })
+                .error(function() {
+                    alert('Unexpected error');
+                });
+            });
+			$( document ).ready(function() {
```
Comment (flaviocopes) [inline admin/templates/guestbook.html.twig]: Can you adjust the indentation here?

File: guestbook.php (hunk 1 above)
Comment (mikegcox) [reply guestbook.php]: This was done by me so that the guestbook entries would show in the Data plugin in Admin.  I forgot to change it back before zipping my Grav folder and giving it to Dalethium.  It should not have been changed.

File: admin/templates/guestbook.html.twig (hunk 2 above)
Comment (Dalethium) [reply admin/templates/guestbook.html.twig]: Sure 1 sec

File: guestbook.php (hunk 1 above)
Comment (Dalethium) [reply guestbook.php]: should it be changed back?

**Comment by flaviocopes** [comment]
Thanks! I'll check the functionality tomorrow. Thanks @mikegcox too, we welcome this kind of contributions.

@Dalethium can you enable the option to allow edits from maintainers? https://github.com/blog/2247-improving-collaboration-with-forks so I can push directly to this PR if needed.

**Comment by Dalethium** [comment]
I think it is on! I can directly add you as a contributor if you wish :)

**Comment by flaviocopes** [comment]
Ok thanks!

**Comment by flaviocopes** [comment]
Functionality-wise, this is going to force moderation. I think we have to add an option in the Plugin's settings and the default should be to not moderate, to preserve the current functionality. Also, I'd rather have a "your message is awaiting moderation" and not list any non-moderated comments, if moderation is on.

Also, currently I could not see the message after submit, but I see it being counted: 

screen shot 2016-10-27 at 12 03 44

and Admin needs some better UI, I see this

screen shot 2016-10-27 at 12 05 50

It could take inspiration from Comments, which has this interface:

screen shot 2016-10-27 at 12 05 44

**Comment by Dalethium** [comment]
on it now

**Comment by Dalethium** [comment]
hopefully that will meet your standards :)

**Comment by Dalethium** [comment]
What do you think of the page buttons? I can remove them if needed, I just figured it'd be marginally better than load more. The most recent will still be at the front, and they won't have to scroll through hell if they have hundreds of comments.

Side note: Approve all can be used to approve all comments from a outdated yaml format.

**Comment by Dalethium** [comment]
the UI looks like this for me now:

image

**Comment by mikegcox** [comment]
Here is my opinion... The pager is a nice addition but I am not sure it is necessary.  I think it would be seldom that a person would need to go through existing messages.  For moderation, all you would ever need to see is the latest entries that have not been approved.  If we keep the pager, the buttons should be made more elegant perhaps like Antimatter but there needs to be padding for sure.  

I also think the pager or load more button should be at the bottom.  It could be fixed to the bottom of the page so it does not float if floating is undesired.

The "Approve all Messages" button also needs padding.

Dalethium mentioned that front-end stuff is not his best skill-set so I don't mind taking a stab at it but I will wait until I get other opinions on the pagers and button styles.

**Comment by Dalethium** [comment]
I still have work to do, but I have made some changes at the request of @mikegcox  -

**Comment by mikegcox** [comment]
screen shot 2016-10-30 at 6 31 28 pm

I did some work on formatting the guestbook admin page.  @flaviocopes, could you please let me know if you like the look of the screen-shot?  We used the same layout as the Data Manager plugin and I styled it using classes that I found on the dashboard.  I think this is very consistent with other Grav plugins.  It is responsive, the table layout used in the comments plugin is not so I like this option a lot more.

**Comment by w00fz** [comment]
Very good job on this, it looks great!

**Comment by flaviocopes** [comment]
Did you add some CSS that is not included? I see it differently: 
screen shot 2016-10-31 at 13 29 32

**Comment by flaviocopes** [comment]
BTW thanks for your effort in improving it :)

**Comment by mikegcox** [comment]
He has not incorporated my styling changes yet.  I tried to fork it but didn't have permission.  We are working together on this and I keep asking for changes so he was getting tired yesterday.  I can add my changes, it is only to the twig file and language file.  I put a title at the top now too so it looks like this now:
screen shot 2016-10-31 at 8 34 18 am

**Comment by mikegcox** [comment]
This is the version with my styling changes.  He has an optional table layout in the code too but I asked hime to remove it as it is not necessary.  I also found a bug that needs to be fixed.  I also requested that all existing messages remain visible on the front-end upon upgrade but I have not tested that yet.

[Mike-changes-guestbook.zip](https://github.com/getgrav/grav-plugin-guestbook/files/561772/Mike-changes-guestbook.zip)

**Comment by Dalethium** [comment]
Sorry for the late changes. I fixed a few bugs with removing deleted posts, I added original functionality as a config option, and I added the UI changes @mikegcox did with a small spacing change on the EM css

**Comment by Dalethium** [comment]
image
This is what it looks like for me

**Comment by mikegcox** [comment]
Thank you for your patience @Dalethium.  I think we have this working very well now.  It is now simplified and nicely styled.  
- When you first upgrade the plugin, the guestbook page of the website will not be affected at all. 
- When you go into admin, the exiting data file will be updated adding a guid and a moderated field for each entry and the site will remain un affected.
- when loaded in Admin, all un-moderated entries show up
- there is a button to show all moderated entries.
- When a delete button is pressed, the entry is removed 
- when the approve button is pressed the delete button goes away
- styling is accomplished using existing css classes from the Admin plugin except for `.center`.  That was added to a `<style>` block.
- the layout is responsive and looks very much like other plugins to Admin

I hope these changes can be incorporated into Guestbook.  The Guestbook plugin is a nice addition to Grav, but needs moderation capabilities.  Even my staging site was getting spammed.  We could tackle the Comments plugin next.

**Comment by mikegcox** [comment]
screen shot 2016-11-01 at 4 15 17 pm

The latest look

**Comment by flaviocopes** [comment]
Thanks, I tested and it's good for me! I will merge _but_ I will add some code to make this an option, which means that moderation is optional (disabled by default) and that the Guestbook sidebar menu will only appear if it's enabled.

P.S. @mikegcox to avoid spam you can add a captcha to the guestbook / comment form

**Review by w00fz — APPROVED** [review]
I was going to suggest the optional factor. Other than that I'm approving as well!

**Comment by mikegcox** [comment]
This is great news.  Thanks guys! 

@flaviocopes, I do have the captcha turned on and it works fine, but without the ability to moderate for the messages there are still undesirable messages being posted and they are live on the site by default.  So if someone has a beef with the site owner, they can voice it and it goes live.  If someone signs the guestbook and they have some spelling and grammar issues, a moderator can delete it from Admin and then repost from the website.  If it were just me moderating on my own site, this change would not have been a show-stopper, I could edit with cpanel or github or even FTP but for a non-technical person, it has to be moderated from the Admin plugin.

**Comment by mikegcox** [comment]
@Dalethium, thanks for your patience... I know it was frustrating at times, but look what we can do now :)

**Comment by flaviocopes** [comment]
@mikegcox yes, it is just a suggestion to avoid automated messages coming in 👍 

@Dalethium @mikegcox thanks for the contribution 😃

**Review by flaviocopes — APPROVED** [review]

**Comment by flaviocopes** [comment]
I just implemented moderation, if disabled in the plugin's settings, the admin panel disappears and  all messages are shown in the frontend.

I still get this issue which I need to look into, do you get the same? It's displaying  `Showing 3 messages of 3 (Note that some may be unmoderated)` but no messages appear, because moderation is enabled but no messages are approved.

**Comment by flaviocopes** [comment]
Tracking in https://github.com/getgrav/grav-plugin-guestbook/issues/8
